core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -218,11 +218,13 @@ class KafkaServer(val config: KafkaConfig, time: Time = Time.SYSTEM, threadNameP
         this.logIdent = logContext.logPrefix
 
         /* create and configure metrics */
-        val reporters = config.getConfiguredInstances(KafkaConfig.MetricReporterClassesProp, classOf[MetricsReporter],
-            Map[String, AnyRef](KafkaConfig.BrokerIdProp -> (config.brokerId.toString)).asJava)
+        val brokerMetricsReporters = new DynamicMetricsReporters(brokerId, config.dynamicConfig)
+        val reporters = new util.ArrayList[MetricsReporter]
+        reporters.addAll(brokerMetricsReporters.currentMetricsReporters.asJava)
         reporters.add(new JmxReporter(jmxPrefix))
         val metricConfig = KafkaServer.metricConfig(config)
         metrics = new Metrics(metricConfig, reporters, time, true)
+        brokerMetricsReporters.initMetrics(metrics)
```

Comment:
This is a little awkward. Couldn't we create the `Metrics` instance with an empty list, then construct the `DynamicMetricsReporters` object and add the reporters?